### ISSUE TITLE
macOS: Cannot add to X509Store [regression from 3.1]

### ISSUE THREAD

**Issue by vcsjones** [dotnet/runtime#39603]
I cannot seem to write to an `X509Store` in macOS. Code to reproduce:

```csharp
using System.Security.Cryptography.X509Certificates;

namespace scratch {
    class Program {
        static void Main(string[] args) {
            using X509Certificate2 cert = new X509Certificate2("cert.pfx", "test", X509KeyStorageFlags.Exportable);
            using X509Store store = new X509Store(StoreName.My, StoreLocation.CurrentUser);
            store.Open(OpenFlags.ReadWrite);
            store.Add(cert);
        }
    }
}
```

This seems to always fail on `store.Add(cert)`. Exception:

```
Unhandled exception. Interop+AppleCrypto+AppleCommonCryptoCryptographicException: UNIX[Undefined error: 0]
   at Interop.AppleCrypto.X509StoreAddCertificate(SafeKeychainItemHandle certOrIdentity, SafeKeychainHandle keychain)
   at Internal.Cryptography.Pal.StorePal.AppleKeychainStore.Add(ICertificatePal cert)
   at System.Security.Cryptography.X509Certificates.X509Store.Add(X509Certificate2 certificate)
   at scratch.Program.Main(String[] args) in /Users/kjones/Projects/scratch/Program.cs:line 9
```

The unit tests for this appear to be in `X509StoreMutableTests_OSX`. When I attempt to run them, all of them are skipped because the "can I write to the store?" conditional fails:

> Writing precondition failed with kPOSIXErrorBase, skipping tests.

This is not an SSH session, or a VM, or anything out of the ordinary, it's a login zsh shell.

**Comment by vcsjones** [comment]
Some other background on this, removing from the keychain does work, so it seems not related to the keychain being locked or in an inconsistent state. I did note that Big Sur is getting changes to how importing certificates _as root_ is handled, but saw nothing describing this exact use case.

Other thing to note: I can't reproduce this in netcoreapp3.1. I wonder if this has anything to do with how the binaries are signed and/or if 5 previews are signed differently than 3.1 official builds. Or, something broke between 3.1 and 5.0.

**Comment by jeffhandley** [comment]
Thank you, @vcsjones! Are you planning to investigate this further? If not, we'll get someone from the team assigned to it so we can get it into 5.0 (probably RC2).

**Comment by vcsjones** [comment]
@jeffhandley 

> Are you planning to investigate this further?

Yeah, but I haven't made a ton of progress. I think a good first step would be to see if someone other than me can repro it, and do the unit tests actually run (not just skipped). Maybe it just doesn't like me.

I can still repro it on Big Sur beta 4, I'm installing 5 right now.

**Comment by wfurt** [comment]
I'm wondering if some of this may depend on system integrity setting. I have 11.0 system up and I can take a look. 
There was at least one other case where 11.0 broke binary compatibility with 10.x. That _could_  explain `kPOSIXErrorBase`.

Running `log  stream  --process <PID_OF_SECURITYD>` may give some more insight @vcsjones.

**Comment by vcsjones** [comment]
> Running log stream --process <PID_OF_SECURITYD> may give some more insight @vcsjones.

Yeah, that's where my stack in https://github.com/dotnet/runtime/issues/39603#issuecomment-678681034 came from.

> I'm wondering if some of this may depend on system integrity setting

Maybe? I'm just thrown by the fact that 3.1 works.

**Comment by wfurt** [comment]
I have repro using latest daily 5.0 build:
```
bash-3.2$ ~/dotnet/dotnet run
Unhandled exception. Interop+AppleCrypto+AppleCommonCryptoCryptographicException: UNIX[Undefined error: 0]
   at Interop.AppleCrypto.X509StoreAddCertificate(SafeKeychainItemHandle certOrIdentity, SafeKeychainHandle keychain)
   at Internal.Cryptography.Pal.StorePal.AppleKeychainStore.Add(ICertificatePal cert)
   at System.Security.Cryptography.X509Certificates.X509Store.Add(X509Certificate2 certificate)
   at scratch.Program.Main(String[] args) in /Users/furt/certs/Program.cs:line 9
```

I will take a look and share findings.

**Comment by wfurt** [comment]
It seems like this is somehow related to deleting temporary keychain _before_ finishing the import @bartonjs. If I comment out that code, (turning AppleCryptoNative_SecKeychainDelete to NOP) import works as expected. 
Does it ring any bell? 

I also verified that 3.1.6 works. I don't have 3.1 build on my 11.0 machine so I did not check if the sequence is the same.

**Comment by vcsjones** [comment]
> deleting temporary keychain before finishing the import

Interesting. Thanks for getting to the bottom of that.

**Comment by bartonjs** [comment]
> Does it ring any bell?

Not particularly... but that must mean we changed the managed code for it... possibly as part of the PFX rewrite?  But I'm surprised an order of operations would have changed.

**Comment by vcsjones** [comment]
I wonder if #39086 is the same root cause.

**Comment by wfurt** [comment]
I tried to do binary search and even if I go back to December/19, adding to store is still failing. So perhaps this is something prior that or it is related to some other build changes.

**Comment by vcsjones** [comment]
The problem is basically here:

https://github.com/dotnet/runtime/blob/afa4b2f344192c3609dd3c1419f68d99d10d7d62/src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/AppleCertificatePal.cs#L106-L113

We import the X509Certificate2 in to a temporary keychain (because `PersistKeySet` is not set). I think the temporary keychain that the key gets imported in to needs to live as long as the returned `ICertificatePal`.

**Comment by wfurt** [comment]
Did that part changed? I thought I saw similar code in 3.1 but I may be mistaken. 
Also, log seems to complain about old-style keychain outside of ~/Library/Keychains. 
I was wondering if the location also may make difference of there is some other way how to avoid that warning. 
(But I did not have time (yet?) for experiments)

**Comment by vcsjones** [comment]
As a test, if that code is changed to:

```c#
SafeKeychainHandle keychain = persist
    ? Interop.AppleCrypto.SecKeychainCopyDefault()
    : Interop.AppleCrypto.CreateTemporaryKeychain();

System.GC.SuppressFinalize(keychain);
```

Then the problem goes away, too.

**Comment by wfurt** [comment]
It seems like holding ref to the keychain is no longer good enough.

**Comment by vcsjones** [comment]
> if the location also may make difference

It doesn't. I tried. (I also tried changing the keychain temp file extension to `keychain-db` in case for some reason that matter)

> Did that part changed?

I don't think so. I still haven't been able to nail down why 3.1 is working, but I thought I might be better off looking in to 5.0. It's entirely possible 3.1 is working by luck.

**Comment by jeffhandley** [comment]
Is there a possibility of a GC change during 5.0 that could have affected this?

**Comment by jeffschwMSFT** [comment]
I asked @JulieLeeMSFT and @mangod9 to take a look.  Sometimes things that have have lifetime issues can 'work' for many years before a slight change in allocations that yield the failure - we often see this with Timers, etc.

**Comment by jeffhandley** [comment]
Thanks, @jeffschwMSFT!

Even if it was introduced by outside factors, I think the fix would likely be along the lines of what @vcsjones [illustrated](https://github.com/dotnet/runtime/issues/39603#issuecomment-681050675).  Are there any objections to moving forward with that as the fix?

**Comment by vcsjones** [comment]
> Are there any objections to moving forward with that as the fix?

That wasn't a complete fix unfortunately, all it did was stop the temp .keychain file from being deleted. That change as-is will leave the disk cluttered with temporary keychains.

I think a correct fix would be to transfer ownership of the `SafeTemporaryKeychainHandle` to the `AppleCertificatePal` iif the PKCS12 file is loaded as non-persisted and the PAL is constructed with a `SafeSecIdentityHandle`. When the PAL is disposed, _then_ we can clean up the temp keychain.

I can give that a shot.

**Comment by jeffschwMSFT** [comment]
I think we need to take a closer look.  Based on the code pattern (embedded within a using), I am surprised that this is failing out of scope.

@jkotas @davidwrighton

**Comment by vcsjones** [comment]
I don't think this is a GC issue - I threw in the SuppressFinalize so the GC wouldn't clean up the handle (which deletes the temp keychain) and force it to leak. That snippet was solely there to at least confirm that is the source of some object-lifetime issue.

**Comment by jeffhandley** [comment]
Got it; thanks for the clarification, @vcsjones.

**Comment by AndyAyersMS** [comment]
I'll take a look at the codegen side to see what we're doing for GC tracking.

**Comment by jkotas** [comment]
> , I am surprised that this is failing out of scope.

Note that this is all new code

**Comment by AndyAyersMS** [comment]
Since this fails with debug codegen it's also unlikely to be a jit GC tracking issue. 

GC info leading up to the call to `ImportPkcs12` looks reasonable:
```asm
000122 E829FBFFFF           call     AppleCrypto:CreateTemporaryKeychain():System.Security.Cryptography.Apple.SafeTemporaryKeychainHandle
000127 48894590             mov      gword ptr [rbp-70H], rax
00012B EB09                 jmp      SHORT G_M61524_IG08
						;; bbWeight=1    PerfScore 13.50
G_M61524_IG07:
00012D E8CEFAFFFF           call     AppleCrypto:SecKeychainCopyDefault():System.Security.Cryptography.Apple.SafeKeychainHandle
000132 48894590             mov      gword ptr [rbp-70H], rax
						;; bbWeight=1    PerfScore 2.00
G_M61524_IG08:
000136 488B7D90             mov      rdi, gword ptr [rbp-70H]
00013A 48897DC0             mov      gword ptr [rbp-40H], rdi
00013E 488B7DC0             mov      rdi, gword ptr [rbp-40H]
000142 48897DB8             mov      gword ptr [rbp-48H], rdi
						;; bbWeight=1    PerfScore 4.00
G_M61524_IG09:
000146 488B7DF0             mov      rdi, bword ptr [rbp-10H]
00014A 488B75F8             mov      rsi, qword ptr [rbp-08H]
00014E 488B55E8             mov      rdx, gword ptr [rbp-18H]
000152 8B4DCC               mov      ecx, dword ptr [rbp-34H]
000155 4C8B45C0             mov      r8, gword ptr [rbp-40H]
000159 E85AC0FFFF           call     Internal.Cryptography.Pal.AppleCertificatePal:ImportPkcs12(System.ReadOnlySpan`1[Byte],Microsoft.Win32.SafeHandles.SafePasswordHandle,bool,System.Security.Cryptography.Apple.SafeKeychainHandle):Internal.Cryptography.ICertificatePal
00015E 488945B0             mov      gword ptr [rbp-50H], rax

;; stack slots rbp-70, rbp-40, rbp-48 are untracked

Register slot id for reg rax = 17.
Register slot id for reg rsi = 18.
Register slot id for reg rdi = 19.
Register slot id for reg rdx = 20.
Register slot id for reg r8 = 21.
Register slot id for reg rcx = 22.

Set state of slot 17 at instr offset 0x127 to Live.
Set state of slot 17 at instr offset 0x12d to Dead.
Set state of slot 17 at instr offset 0x132 to Live.
Set state of slot 17 at instr offset 0x136 to Dead.
Set state of slot 19 at instr offset 0x13a to Live.
Set state of slot 19 at instr offset 0x146 to Dead.
Set state of slot 16 at instr offset 0x14a to Live.
Set state of slot 20 at instr offset 0x152 to Live.
Set state of slot 21 at instr offset 0x159 to Live.
Set state of slot 17 at instr offset 0x15e to Live.

```

**Comment by jeffhandley** [comment]
Hey @vcsjones -- @wfurt is going to take a look into this this week.  Thanks for your help on the investigation; you've given us a great place to continue from. :100:

**Comment by jeffhandley** [comment]
I'm editing the title to remove "Big Sur" since this repro'd for me on Catalina as well.

**Comment by vcsjones** [comment]
> @wfurt is going to take a look into this this week.

Thanks a ton. Sorry I've just been underwater with other things lately.

**Comment by wfurt** [comment]
reopening to 5.0. Master is now for 6.0

**Comment by danmosemsft** [comment]
I was wondering what regressed it, and I guess it was probably the big PFX reader/writer change: https://github.com/dotnet/runtime/commit/af6ab5f364871cb3f5d301d5f9c6b786bccb8a89

**Comment by jeffhandley** [comment]
Fixed in 5.0.0 as well with #41882.
